File: src/types/evidence.ts
export type EvidenceSourceType =
  | 'FIR'
  | 'POLICE_REPORT'
  | 'CDR'
  | 'FINANCIAL_RECORD'
  | 'SURVEILLANCE_REPORT'
  | 'INTELLIGENCE_REPORT'
  | 'SOCIAL_MEDIA_EXPORT'
  | 'CRIMINAL_HISTORY'
  | 'VEHICLE_RECORD'
  | 'IDENTITY_DOCUMENT'
  | 'CCTV_VIDEO'
  | 'CCTV_SNAPSHOT'
  | 'AUDIO'
  | 'IMAGE'
  | 'OTHER';

export type ProcessingStatus =
  | 'UPLOADED'
  | 'VALIDATING'
  | 'QUEUED'
  | 'PROCESSING'
  | 'PROCESSED'
  | 'PARTIALLY_PROCESSED'
  | 'FAILED'
  | 'QUARANTINED';

export type Classification = 'PUBLIC' | 'INTERNAL' | 'RESTRICTED' | 'CONFIDENTIAL';

export type JobType =
  | 'FILE_VALIDATION'
  | 'TEXT_EXTRACTION'
  | 'OCR'
  | 'ENTITY_EXTRACTION'
  | 'CDR_INGESTION'
  | 'FINANCIAL_INGESTION'
  | 'VIDEO_ANALYSIS'
  | 'REPORT_GENERATION';

export type JobStatus = 'QUEUED' | 'RUNNING' | 'COMPLETED' | 'FAILED' | 'CANCELLED';

export interface EvidenceFile {
  id: string;
  original_filename: string;
  stored_filename: string;
  mime_type: string;
  extension: string;
  size_bytes: number;
  storage_provider: string;
  storage_key: string;
  checksum_sha256: string | null;
  uploaded_by: string;
  uploaded_at: string;
  updated_at: string;
  created_at: string;
  source_type: EvidenceSourceType;
  classification: Classification;
  description: string | null;
  tags: string[];
  case_id: string | null;
  investigation_id: string | null;
  processing_status: ProcessingStatus;
  processing_error: string | null;
  processed_at: string | null;
  version: number;
  parent_file_id: string | null;
  is_original: boolean;
  is_deleted: boolean;
}

export interface EvidenceVersion {
  id: string;
  file_id: string;
  version: number;
  stored_filename: string;
  storage_key: string;
  size_bytes: number;
  checksum_sha256: string | null;
  uploaded_by: string;
  uploaded_at: string;
  is_original: boolean;
  change_description: string | null;
}

export interface EvidenceAudit {
  id: string;
  file_id: string;
  actor_user_id: string;
  action: string;
  resource_type: string;
  resource_id: string;
  timestamp: string;
  metadata: Record<string, any>;
}

export interface ProcessingJob {
  id: string;
  file_id: string;
  job_type: JobType;
  status: JobStatus;
  progress: number;
  error: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_by: string;
  metadata: Record<string, any>;
  created_at: string;
}

export interface IngestionResult {
  id: string;
  file_id: string;
  job_id: string | null;
  total_records: number;
  valid_records: number;
  invalid_records: number;
  duplicate_records: number;
  warnings: string[] | null;
  errors: string[] | null;
  metadata: Record<string, any>;
  created_at: string;
}

export interface ExtractedEntity {
  id: string;
  file_id: string;
  job_id: string | null;
  entity_type: string;
  canonical_name: string;
  aliases: string[] | null;
  attributes: Record<string, any> | null;
  confidence: number;
  extraction_method: string;
  source_page: number | null;
  source_text: string | null;
  created_at: string;
}

export interface ExtractedRelationship {
  id: string;
  file_id: string;
  job_id: string | null;
  source_entity_id: string | null;
  target_entity_id: string | null;
  relationship_type: string;
  weight: number;
  confidence: number;
  source_record_id: string | null;
  metadata: Record<string, any> | null;
  created_at: string;
}

export type CaseStatus = 'OPEN' | 'UNDER_INVESTIGATION' | 'ON_HOLD' | 'CLOSED' | 'ARCHIVED';
export type CasePriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface Case {
  id: string;
  case_number: string | null;
  title: string;
  description: string;
  status: CaseStatus;
  priority: CasePriority;
  classification: Classification;
  location: string | null;
  tags: string[] | null;
  assigned_investigators: string[] | null;
  related_entities: Array<{ type: string; id: string }> | null;
  related_evidence: string[] | null;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/evidence.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, Dict, Any, List
from datetime import datetime

# --- Enums / Constants ---
EVIDENCE_SOURCE_TYPES = [
    "FIR",
    "POLICE_REPORT",
    "CDR",
    "FINANCIAL_RECORD",
    "SURVEILLANCE_REPORT",
    "INTELLIGENCE_REPORT",
    "SOCIAL_MEDIA_EXPORT",
    "CRIMINAL_HISTORY",
    "VEHICLE_RECORD",
    "IDENTITY_DOCUMENT",
    "CCTV_VIDEO",
    "CCTV_SNAPSHOT",
    "AUDIO",
    "IMAGE",
    "OTHER",
]

PROCESSING_STATUSES = [
    "UPLOADED",
    "VALIDATING",
    "QUEUED",
    "PROCESSING",
    "PROCESSED",
    "PARTIALLY_PROCESSED",
    "FAILED",
    "QUARANTINED",
]

CLASSIFICATIONS = ["PUBLIC", "INTERNAL", "RESTRICTED", "CONFIDENTIAL"]


# --- Evidence File ---
class EvidenceFileBase(BaseModel):
    original_filename: str
    mime_type: str
    extension: str
    size_bytes: int
    storage_provider: str = "local"
    storage_key: str
    checksum_sha256: Optional[str] = None

    source_type: str = "OTHER"
    classification: str = "INTERNAL"
    description: Optional[str] = None
    tags: Optional[List[str]] = None

    case_id: Optional[str] = None
    investigation_id: Optional[str] = None

    processing_status: str = "UPLOADED"
    processing_error: Optional[str] = None
    processed_at: Optional[datetime] = None

    version: int = 1
    parent_file_id: Optional[str] = None
    is_original: bool = True
    is_deleted: bool = False


class EvidenceFileCreate(EvidenceFileBase):
    uploaded_by: str


class EvidenceFileUpdate(BaseModel):
    description: Optional[str] = None
    tags: Optional[List[str]] = None
    case_id: Optional[str] = None
    investigation_id: Optional[str] = None
    classification: Optional[str] = None
    processing_status: Optional[str] = None
    processing_error: Optional[str] = None
    processed_at: Optional[datetime] = None
    is_deleted: Optional[bool] = None


class EvidenceFileResponse(EvidenceFileBase):
    id: str
    uploaded_by: str
    uploaded_at: datetime
    updated_at: datetime
    created_at: datetime


# --- Evidence Version ---
class EvidenceVersionBase(BaseModel):
    file_id: str
    version: int
    stored_filename: str
    storage_key: str
    size_bytes: int
    checksum_sha256: Optional[str] = None
    uploaded_by: str
    uploaded_at: datetime
    is_original: bool = False
    change_description: Optional[str] = None


class EvidenceVersionResponse(EvidenceVersionBase):
    id: str


# --- Evidence Audit ---
class EvidenceAuditBase(BaseModel):
    file_id: str
    actor_user_id: str
    action: str
    resource_type: str = "evidence_file"
    resource_id: str
    metadata: Optional[Dict[str, Any]] = None


class EvidenceAuditResponse(EvidenceAuditBase):
    id: str
    timestamp: datetime


# --- Processing Job ---
class ProcessingJobBase(BaseModel):
    file_id: str
    job_type: str
    status: str = "QUEUED"
    progress: float = 0.0
    error: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_by: str
    metadata: Optional[Dict[str, Any]] = None


class ProcessingJobCreate(ProcessingJobBase):
    pass


class ProcessingJobUpdate(BaseModel):
    status: Optional[str] = None
    progress: Optional[float] = None
    error: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    metadata: Optional[Dict[str, Any]] = None


class ProcessingJobResponse(ProcessingJobBase):
    id: str
    created_at: datetime


# --- Ingestion Result (CDR / Financial) ---
class IngestionResultBase(BaseModel):
    file_id: str
    job_id: Optional[str] = None
    total_records: int = 0
    valid_records: int = 0
    invalid_records: int = 0
    duplicate_records: int = 0
    warnings: Optional[List[str]] = None
    errors: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None


class IngestionResultCreate(IngestionResultBase):
    pass


class IngestionResultResponse(IngestionResultBase):
    id: str
    created_at: datetime


# --- Entity Extraction Result ---
class ExtractedEntityBase(BaseModel):
    file_id: str
    job_id: Optional[str] = None
    entity_type: str
    canonical_name: str
    aliases: Optional[List[str]] = None
    attributes: Optional[Dict[str, Any]] = None
    confidence: float = 0.0
    extraction_method: str = "regex"
    source_page: Optional[int] = None
    source_text: Optional[str] = None


class ExtractedEntityResponse(ExtractedEntityBase):
    id: str
    created_at: datetime


class ExtractedRelationshipBase(BaseModel):
    file_id: str
    job_id: Optional[str] = None
    source_entity_id: Optional[str] = None
    target_entity_id: Optional[str] = None
    relationship_type: str
    weight: float = 0.5
    confidence: float = 0.0
    source_record_id: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class ExtractedRelationshipResponse(ExtractedRelationshipBase):
    id: str
    created_at: datetime
